optimum/onnxruntime/modeling_diffusion.py

```diff
@@ -465,35 +538,133 @@ def dtype(self):
 
         return None
 
-    def to(self, *args, device: Optional[Union[torch.device, str, int]] = None, dtype: Optional[torch.dtype] = None):
+    def to(self, *args, device: Optional[Union[int, str, torch.device]] = None, dtype: Optional[torch.dtype] = None):
         for arg in args:
-            if isinstance(arg, torch.device):
-                device = arg
-            elif isinstance(arg, (int, str)):
+            if isinstance(arg, (int, str, torch.device)):
                 device = torch.device(arg)
             elif isinstance(arg, torch.dtype):
                 dtype = arg
 
-        if device is not None and device != self.device:
-            raise ValueError(
-                "Cannot change the device of a pipeline part without changing the device of the parent pipeline. "
-                "Please use the `to` method of the parent pipeline to change the device."
-            )
-
         if dtype is not None and dtype != self.dtype:
             raise NotImplementedError(
                 f"Cannot change the dtype of the pipeline from {self.dtype} to {dtype}. "
                 f"Please export the pipeline with the desired dtype."
             )
 
-    def prepare_onnx_inputs(self, use_torch: bool, **inputs: Union[torch.Tensor, np.ndarray]) -> Dict[str, np.ndarray]:
+        if device is None or device == self.device:
+            return self
+
+        device, provider_options = parse_device(device)
+        provider = get_provider_for_device(device)
+        validate_provider_availability(provider)
+
+        self.session.set_providers([provider], provider_options=[provider_options])
+
+        self._providers = self.session.get_providers()
+        self._providers_options = self.session.get_provider_options()
+
+        if self.provider != provider or self.provider_options != provider_options:
+            raise ValueError(
+                f"Failed to set the device to {device}. "
+                f"Requested provider {provider} with options: {provider_options}, "
+                f"but got provider {self.provider} with options: {self.provider_options}."
+            )
+
+        self._device = device
+
+        return self
+
+    def _get_output_shapes(self, **model_inputs: torch.Tensor) -> Dict[str, int]:
```

Comment:
This function is very slow (in milliseconds). Any way to improve the performance?